### PR TITLE
feat: runbooks queried from GitHub and Gitlab

### LINKED ISSUE #1780: [➕ Feature]: Runbooks queried by GitHub and GitLab providers
We want to introduce "Runbooks" as a content queried from GitHub and GitLab providers. For the beginning, Runbook in Keep will be a "title" and a link to the ".md" file.

# Update providers

Introduce "Runbooks" as data type providers could query from the source. Similar to "alerts", or "topology".
Screenshot 2024-09-01 at 18 32 29

For GitHub and Gitlab, I think the "Optional settings" (similar to prometheus) should allow choosing the repo and the path with ".md" files.
Screenshot 2024-09-01 at 18 32 16

# Runbooks page

Should represent all runbooks and Incidents those Runbooks are attached to. Just a table.

# Update the Incident page

Allow attaching/unattaching of runbooks to the incident.

### PR DESCRIPTION
**Pull request by Mubashirshariq**
Closes #1780 
/claim #1780 

## ✅ Checks

- [ ] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

### CODE REVIEW

**Comment by Mubashirshariq** [comment]
@Matvey-Kuk  please review

File: keep/runbooks/runbooks_service.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            session.commit()
+            session.expire(new_runbook, ["contents"])
+            session.refresh(new_runbook)  # Refresh the runbook instance
+            result = RunbookDtoOut.from_orm(new_runbook)
+            return result
+        except ValidationError as e:
+            logger.exception(f"Failed to create runbook {e}")
+
+    @staticmethod
+    def get_all_runbooks(session: Session, tenant_id: str, limit=25, offset=0) -> dict:
```
Comment (Matvey-Kuk) [inline keep/runbooks/runbooks_service.py:52]: This file content sounds like something belonging to keep.api.core.db.py, what do you think?

File: keep/api/models/db/migrations/versions/2024-10-08-22-23_0796be640663.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+"""Add  runbook tables
```
Comment (Matvey-Kuk) [inline keep/api/models/db/migrations/versions/2024-10-08-22-23_0796be640663.py:1]: I would appreciate if both migrations will be merged together.

File: keep-ui/app/runbooks/runbook-table.tsx [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+              contents.length - 4
+            } more...`}</Badge>
+          )}
+        </div>
+      );
+    },
+  }),
+] as DisplayColumnDef<RunbookV2>[];
+
+// function PreviewContent({type, data}:{type:string, data:string}){
```
Comment (Matvey-Kuk) [inline keep-ui/app/runbooks/runbook-table.tsx:96]: Please remove commented-out code.

**Review by Matvey-Kuk — CHANGES_REQUESTED** [review]
A few minor changes, in general looks good! Will test locally on the next iteration

File: keep-ui/app/runbooks/runbook-table.tsx (hunk 3 above)
Comment (rajeshj11) [reply keep-ui/app/runbooks/runbook-table.tsx:96]: ack

File: keep/api/models/db/migrations/versions/2024-10-08-22-23_0796be640663.py (hunk 2 above)
Comment (rajeshj11) [reply keep/api/models/db/migrations/versions/2024-10-08-22-23_0796be640663.py:1]: ack

File: keep/runbooks/runbooks_service.py (hunk 1 above)
Comment (rajeshj11) [reply keep/runbooks/runbooks_service.py:52]: Yes, that makes sense. I referenced the topology feature.(so followed the same pattren)

**Comment by rajeshj11** [comment]
@Matvey-Kuk @talboren  I wanted to follow up on this PR, as it seems to have been pending for a while.

**Review by Matvey-Kuk — CHANGES_REQUESTED** [review]
@Mubashirshariq thank you for the PR! I have a few comments:

1) I think there is some confusion between Runbook & Runbook Path. For now, table is representing Runbook Folders, but not actual Runbooks which are individual .md files.
2) I didn't find a way to link individual Runbook.md file to an incident, only the whole path.
3) Linked runbook is not visible at the incident page.
4) As well as I understood I could use "Settings" button to query runbooks from GitHub, what do we need "GitHub Repository Name" & "Path to .md files in the repository" at the provider settings page? Sounds like a duplicate for me.

Screenshot 2024-10-20 at 13 47 13


P.S. Fixed the migration.

**Comment by Matvey-Kuk** [comment]
Plz merge this https://github.com/keephq/keep/commit/84cf08d82f88ebd5120df8367e04994f5bfbaefb to your branch ;)

**Comment by rajeshj11** [comment]
> Plz merge this [84cf08d](https://github.com/keephq/keep/commit/84cf08d82f88ebd5120df8367e04994f5bfbaefb) to your branch ;)

ack

**Comment by rajeshj11** [comment]
> @Mubashirshariq thank you for the PR! I have a few comments:
> 
> 1. I think there is some confusion between Runbook & Runbook Path. For now, table is representing Runbook Folders, but not actual Runbooks which are individual .md files.
> 2. I didn't find a way to link individual Runbook.md file to an incident, only the whole path.
> 3. Linked runbook is not visible at the incident page.
> 4. As well as I understood I could use "Settings" button to query runbooks from GitHub, what do we need "GitHub Repository Name" & "Path to .md files in the repository" at the provider settings page? Sounds like a duplicate for me.
> 
> Screenshot 2024-10-20 at 13 47 13
> P.S. Fixed the migration.

1. I think there is some confusion between Runbook & Runbook Path. For now, table is representing Runbook Folders, but not actual Runbooks which are individual .md files.
Ideally, we should ensure that users provide only one valid markdown file path (folder paths being supported exclusively for GitHub).ex: README.md <realtive_pat_of_md file>

3. Linked runbook is not visible at the incident page.
 I didn’t fully understand this part. Could you provide more context on how we are currently linking or associating the runbook with the incident?
 
 We assumed this flow for linking the runbook to incident.
https://www.loom.com/share/22c5e9924278426fbd8900ea55f545a9

**Comment by Matvey-Kuk** [comment]
> I didn’t fully understand this part. Could you provide more context on how we are currently linking or associating the runbook with the incident?
I quite don't understand the question. The linking between incident and a runbook is expected to be added by this PR. There is many-many link expected. If I as a user link runbook to the incident, I expect to see the link from the incident page side as well.

**Comment by rajeshj11** [comment]
@Matvey-Kuk The way we’ve implemented it—adding this in the config—would require significant effort. Should we create a subtask to address this?"

**Comment by rajeshj11** [comment]
> > I didn’t fully understand this part. Could you provide more context on how we are currently linking or associating the runbook with the incident?
> 
> I quite don't understand the question. The linking between incident and a runbook is expected to be added by this PR. There is many-many link expected. If I as a user link runbook to the incident, I expect to see the link from the incident page side as well.

Screenshot from 2024-10-20 21-01-03
you mean this functionality for runbook too.

**Comment by rajeshj11** [comment]
> > > I didn’t fully understand this part. Could you provide more context on how we are currently linking or associating the runbook with the incident?
> > 
> > 
> > I quite don't understand the question. The linking between incident and a runbook is expected to be added by this PR. There is many-many link expected. If I as a user link runbook to the incident, I expect to see the link from the incident page side as well.
> 
> Screenshot from 2024-10-20 21-01-03 you mean this functionality for runbook too.

^^

**Comment by rajeshj11** [comment]
> > > I didn’t fully understand this part. Could you provide more context on how we are currently linking or associating the runbook with the incident?
> > 
> > 
> > I quite don't understand the question. The linking between incident and a runbook is expected to be added by this PR. There is many-many link expected. If I as a user link runbook to the incident, I expect to see the link from the incident page side as well.
> 
> Screenshot from 2024-10-20 21-01-03 you mean this functionality for runbook too.

@Matvey-Kuk ^^

**Comment by Mubashirshariq** [comment]
> > > I didn’t fully understand this part. Could you provide more context on how we are currently linking or associating the runbook with the incident?
> > 
> > 
> > I quite don't understand the question. The linking between incident and a runbook is expected to be added by this PR. There is many-many link expected. If I as a user link runbook to the incident, I expect to see the link from the incident page side as well.
> 
> Screenshot from 2024-10-20 21-01-03 you mean this functionality for runbook too.


@Matvey-Kuk 
All the other things are done only thing that is pending is incident one could you please provide some clarifications regarding this

https://github.com/user-attachments/assets/6887c145-82a7-44b6-9007-b6773450c260

**Comment by Matvey-Kuk** [comment]
@Mubashirshariq thank you for the contribution! The core of the codebase is at good quality, but the UI and some UX decisions confuse me. I'm going to close the PR and proceed with the bounty payment via your GH sponsorship. @VladimirFilonov will take your codebase and continue working on it in order to bring it up to the UX standards we have.

**Comment by Matvey-Kuk** [comment]
Screenshot 2024-10-29 at 15 53 38

**Pull request closed without merging**